cluster/transport/client.go

```diff
@@ -101,59 +100,67 @@ func (cl *Client) Notify(ctx context.Context, rAddr string, req *cmd.NotifyPeerR
 // Remove removes this node from an existing cluster identified by its leader's address.
 // If a new leader has been elected, the request is redirected to the new leader.
 func (cl *Client) Remove(ctx context.Context, leaderAddress string, req *cmd.RemovePeerRequest) (*cmd.RemovePeerResponse, error) {
-	addr, err := cl.rpc.Address(leaderAddress)
-	if err != nil {
-		return nil, fmt.Errorf("resolve address: %w", err)
-	}
-
-	conn, err := grpc.DialContext(ctx, addr, grpc.WithTransportCredentials(insecure.NewCredentials()))
+	conn, err := cl.getConn(leaderAddress)
 	if err != nil {
-		return nil, fmt.Errorf("dial: %w", err)
+		return nil, err
 	}
-	defer conn.Close()
 	c := cmd.NewClusterServiceClient(conn)
 	return c.RemovePeer(ctx, req)
 }
 
 func (cl *Client) Apply(leaderAddr string, req *cmd.ApplyRequest) (*cmd.ApplyResponse, error) {
 	ctx := context.Background()
-	addr, err := cl.rpc.Address(leaderAddr)
+	conn, err := cl.getConn(leaderAddr)
 	if err != nil {
-		return nil, fmt.Errorf("resolve address: %w", err)
+		return nil, err
 	}
 
-	conn, err := grpc.DialContext(
-		ctx,
-		addr,
-		grpc.WithTransportCredentials(insecure.NewCredentials()),
-		grpc.WithDefaultServiceConfig(serviceConfig),
-	)
-	if err != nil {
-		return nil, fmt.Errorf("dial: %w", err)
-	}
-	defer conn.Close()
 	c := cmd.NewClusterServiceClient(conn)
 	return c.Apply(ctx, req)
 }
 
 func (cl *Client) Query(ctx context.Context, leaderAddress string, req *cmd.QueryRequest) (*cmd.QueryResponse, error) {
+	conn, err := cl.getConn(leaderAddress)
+	if err != nil {
+		return nil, err
+	}
+
+	c := cmd.NewClusterServiceClient(conn)
+	return c.Query(ctx, req)
+}
+
+func (cl *Client) Close() {
+	cl.leaderConn.Close()
+}
+
+func (cl *Client) getConn(leaderAddress string) (*grpc.ClientConn, error) {
+	cl.connLock.Lock()
+	defer cl.connLock.Unlock()
+
+	if cl.leaderConn != nil && leaderAddress == cl.leaderAddr {
+		return cl.leaderConn, nil
+	}
+
+	if cl.leaderConn != nil {
+		// close open conn if leader addr changed
+		cl.leaderConn.Close()
+	}
+
 	addr, err := cl.rpc.Address(leaderAddress)
 	if err != nil {
 		return nil, fmt.Errorf("resolve address: %w", err)
 	}
 
-	conn, err := grpc.DialContext(
-		ctx,
+	conn, err := grpc.Dial(
```

Comment:
question(blocking): Where is `cl.leaderConn` set to conn ? I don't think we're actually reusing the connection with that change

Reply:
🤦🏼  good catch